### PR TITLE
Add preview panel for entries and groups, closes #454 and #558

### LINKED ISSUE #454: Add preview pane
## Expected Behavior
The highlighted entry should have a preview pane with the details of the entry i.e. group, creation, username etc

## Possible Solution
KeepassX 0.4.x had this feature no idea if you can use the old code.

## Context
It's handy to see various info in the pane, especially the clickable urls.

### PR DESCRIPTION
**Pull request by TheZ3ro**
## Description

This PR add a preview panel for entires and groups and add/fix some minor things:
- Fixed Entry.cpp resolveUrl when the url is empty
- Added Group.cpp hierarchy that return a QStringList of parent group names
- Added chronometer icon for TOTP
- Details panel is visible by default (can be hidden permanently in settings, temporary with the slider or the X button) 
- Passwords are hidden by default (can be changed in settings)
- Show Autotype association for entries (suggested by @droidmonkey )
- Protected attributes are shown as `[PROTECTED]`
- Root group isn't shown in entries' path
- Hi-res icons are resized to 16x16 to don't mess up the layout
- Remove Creation/Access/Modification data (suggested by @andusvan)
- Remove Attributes and Autotype Tabs for Groups (they can't have those) (suggested by @andusvan)
- Selected tab isn't changed when selecting a new entry/group (suggested by @andusvan)

Some info on the preview panel:

The preview panel can be disabled (and hidden) in the settings, it can also be hidden using the spacer above it.
It can be temporary hidden using the X icon, the panel will be visible again when a new Entry/Group is clicked.

You must click on an Entry/Group to display its info on the panel, this is due to a bug when using both click and change signals (currently the change signals are commented, line 196-198  `src/gui/DatabaseWidget.cpp`)

You can decide to disable/hide passwords in the preview panel from the settings, in this case the visualized password will be 4 dot. 

## Motivation and context

Closes #454 and #558

## How has this been tested?

Manual gui testing

## Screenshots (if appropriate):
Screenshots here https://github.com/keepassxreboot/keepassxc/issues/454#issuecomment-330925919

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by phoerious** [comment]
Also you may want to adjust the spacing a little. The second and third column are way closer to each other than the first and second and the third and fourth.

**Comment by TheZ3ro** [comment]
@phoerious the spacing is due to the PasswordEdit and the visibility button, there isn't much I can do. 

Using a label for the password will bring the same problem as the URL since most of my password are >40 char and such length will broke the layout on small window size.
Should I short passwords like it's done with URLs?

@droidmonkey like I said in the first message, enabling that will bring some problems with guitest that I can't reproduce manually

**Comment by phoerious** [comment]
Show the password as three asterisks or bullets, don't use the password edit. Add an extra button to reveal it. Maybe add some sort of context pop-up on hover for all entries to show a non-elided full-length version.

**Comment by TheZ3ro** [comment]
I've fixed the navigating issue, the notes label and the password field. Should be good now :wink:

**Comment by droidmonkey** [comment]
I'll take a look tomorrow but there is still considerable whitespace issues (as of the last screenshot) and questionable need to display creation/access times.

**Comment by droidmonkey** [comment]
Looks great! Found an issue though:

xc_hires_icon_cutoff

The new hi-res icons get cut-off and enlarge the view.

Also, do we want to repeat the root group's folder in the path? It's technically not a subfolder and could be confusing.

**Comment by TheZ3ro** [comment]
> The new hi-res icons get cut-off and enlarge the view.

Woops, I didn't tested with hi-res icon

> Also, do we want to repeat the root group's folder in the path?

You mean removing the first `/` ? 

> The password field is also unmasked.

There should be an option in Settings > Security > `Hide passwords in the preview panel`, by default they are always visible

> Suggestion: It might be nice to add an auto-type tab that lists the currently assigned auto-type window/sequence combos for the entry.

Nice, I will work on this

**Comment by droidmonkey** [comment]
I meant removing the "NewDatabase" label

**Comment by TheZ3ro** [comment]
Oh ok got it. I will work on this in the next few days

**Comment by TheZ3ro** [comment]
Rebased on develop, ready for review

File: src/gui/DetailsWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    QStringList customAttributes = m_currentEntry->attributes()->customKeys();
+    if (customAttributes.size() > 0) {
+        m_ui->tabWidget->setTabEnabled(AttributesTab, true);
+        m_ui->attributesEdit->clear();
+
+        QString attributesText = QString();
+        for (const QString& key : customAttributes) {
+            QString value = m_currentEntry->attributes()->value(key);
+            if (m_currentEntry->attributes()->isProtected(key)) {
+                value = "<i>[PROTECTED]</i>";
```
Comment (phoerious) [inline src/gui/DetailsWidget.cpp]: Not translated. Re-use locale string from edit entry screen (without the `i` tags).

File: src/gui/DetailsWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            }
+            attributesText.append(QString("<b>%1</b>: %2<br/>").arg(key, value));
+        }
+        m_ui->attributesEdit->setText(attributesText);
+    }
+
+    m_ui->autotypeTree->clear();
+    AutoTypeAssociations* autotypeAssociations = m_currentEntry->autoTypeAssociations();
+    QList<QTreeWidgetItem *> items;
```
Comment (phoerious) [inline src/gui/DetailsWidget.cpp]: Remove space before *.

File: src/gui/DetailsWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        m_ui->totpWidget->hide();
+    }    
+}
+
+QString DetailsWidget::shortUrl(QString url)
+{
+    QString newurl = "";
+    if (url.length() > 60) {
+        newurl.append(url.left(20));
+        newurl.append("...");
```
Comment (phoerious) [inline src/gui/DetailsWidget.cpp]: This should be an ellipsis character (U+2026).

File: src/gui/DetailsWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    }
+    return url;
+}
+
+QString DetailsWidget::shortPassword(QString password)
+{
+    QString newpassword = "";
+    if (password.length() > 60) {
+        newpassword.append(password.left(50));
+        newpassword.append("...");
```
Comment (phoerious) [inline src/gui/DetailsWidget.cpp]: Same here.

File: src/gui/DetailsWidget.cpp (hunk 1 above)
Comment (TheZ3ro) [reply src/gui/DetailsWidget.cpp]: In the edit entry it's `tr("[PROTECTED] Press reveal to view or edit")`

I will separate the 2 string so we can re-use `[PROTECTED]`

File: src/gui/entry/EditEntryWidget.cpp [hunk 5]
```diff
@@ -632,7 +632,7 @@ void EditEntryWidget::displayAttribute(QModelIndex index, bool showProtected)
     if (index.isValid()) {
         QString key = m_attributesModel->keyByIndex(index);
         if (showProtected) {
-            m_advancedUi->attributesEdit->setPlainText(tr("[PROTECTED] Press reveal to view or edit"));
+            m_advancedUi->attributesEdit->setPlainText(tr("[PROTECTED]")+" "+tr("Press reveal to view or edit"));
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Please add spaces between operators or use .append().

**Comment by TheZ3ro** [comment]
What is missing before merge?

**Comment by droidmonkey** [comment]
Let's merge this, we can make tweaks later as we use it more.

**Comment by dprime** [comment]
I know this is closed, but I've got some feedback on the layout of this. In the old keepass2, I really liked and used a lot the notes fields for storing bits and pieces. Being able to view the notes in the preview panel with 1 click is a really nice feature and something that stops me moving from the old keepass2 to keepassxc. With this implementation, the notes field is hidden by default. Would it not be more usable to display everything in one, like the old keepass2?

**Comment by droidmonkey** [comment]
Notes are visible in a tab, you can select that tab and it will persist between entry selections.

**Pull request merged**